Code/GraphMol/FileParsers/MultithreadedSDMolSupplier.cpp

```diff
@@ -0,0 +1,262 @@
+  df_sanitize = sanitize;
+  df_removeHs = removeHs;
+  df_strictParsing = strictParsing;
+  d_numWriterThreads = numWriterThreads;
+  d_sizeInputQueue = sizeInputQueue;
+  d_sizeOutputQueue = sizeOutputQueue;
+  d_inputQueue =
+      new ConcurrentQueue<std::tuple<std::string, unsigned int, unsigned int>>(
+          d_sizeInputQueue);
+  d_outputQueue =
+      new ConcurrentQueue<std::tuple<ROMol *, std::string, unsigned int>>(
+          d_sizeOutputQueue);
+
+  df_end = false;
+  d_line = 0;
+  df_processPropertyLists = true;
+}
+
+MultithreadedSDMolSupplier::~MultithreadedSDMolSupplier() {
+  if (df_owner && dp_inStream) {
+    delete dp_inStream;
+    df_owner = false;
+    dp_inStream = nullptr;
+  }
+}
+
+// ensures that there is a line available to be read
+// from the file, implementation identical to the method in
+// in ForwardSDMolSupplier
+void MultithreadedSDMolSupplier::checkForEnd() {
+  PRECONDITION(dp_inStream, "no stream");
+  // we will call it end of file if we have more than 4 contiguous empty lines
+  // or we reach end of file in the meantime
+  if (dp_inStream->eof()) {
+    df_end = true;
+    return;
+  }
+
+  // we are not at the end of file, check for blank lines
+  unsigned int numEmpty = 0;
+  std::string tempStr;
+  // in case df_end is not set then, reset file pointer
+  std::streampos holder = dp_inStream->tellg();
+  for (unsigned int i = 0; i < 4; i++) {
+    tempStr = getLine(dp_inStream);
+    if (dp_inStream->eof()) {
+      df_end = true;
+      break;
+    }
+    if (tempStr.find_first_not_of(" \t\r\n") == std::string::npos) {
+      ++numEmpty;
+    }
+  }
+  if (numEmpty == 4) {
+    df_end = true;
+  }
+  // we need to reset the file pointer to read the next record
+  if (!df_end) {
+    dp_inStream->clear();
+    dp_inStream->seekg(holder);
+  }
+}
+
+bool MultithreadedSDMolSupplier::getEnd() const {
+  PRECONDITION(dp_inStream, "no stream");
+  return df_end;
+}
+
+bool MultithreadedSDMolSupplier::extractNextRecord(std::string &record,
+                                                   unsigned int &lineNum,
+                                                   unsigned int &index) {
+  PRECONDITION(dp_inStream, "no stream");
+  if (dp_inStream->eof()) {
+    df_end = true;
+    return false;
+  }
+
+  std::string currentStr, prevStr;
+  record = "";
+  lineNum = d_line;
+  while (!dp_inStream->eof() && !dp_inStream->fail() &&
+         (prevStr.find_first_not_of(" \t\r\n") != std::string::npos ||
+          currentStr[0] != '$' || currentStr.substr(0, 4) != "$$$$")) {
+    prevStr = currentStr;
+    std::getline(*dp_inStream, currentStr);
+    record += currentStr + "\n";
+    ++d_line;
+    if (prevStr.find_first_not_of(" \t\r\n") == std::string::npos &&
+        currentStr[0] == '$' && currentStr.substr(0, 4) == "$$$$") {
+      this->checkForEnd();
+    }
+  }
+  index = d_currentRecordId;
+  ++d_currentRecordId;
+  return true;
+}
+
+void MultithreadedSDMolSupplier::readMolProps(ROMol *mol,
```

Comment:
Instead of copying code in from a method on another class, it would be better to define a helper function and call that from here and `ForwardSDMolSupplier`.

Reply:
The `readMolProps` method implementation is also slightly different since the method in `ForwardSDMolSupplier` uses the `dp_inStream` to read molecule properties, but since we already have the record as a string, I convert the string record into a stream and then use that to read molecule properties. This way we do not disturb the actual file pointers so that the next record can be read correctly.